composer/utils/batch_helpers.py

```diff
@@ -0,0 +1,226 @@
+    if set_fn:
+        return set_fn(batch, value)
+    if isinstance(key, Sequence) and not isinstance(key, str):
+        return _batch_set_multiple(batch, key, value)
+    else:
+        return _batch_set(batch, key, value)
+
+
+def _batch_set(batch: Any, key: Any, value: Any) -> Any:
+    """Sets a key value pair in a non-tuple batch."""
+    if isinstance(batch, tuple):
+        return _batch_set_tuple(batch, key, value)
+
+    exceptions = []
+    try:
+        # Check if one can do a __getitem__ before doing a __setitem__ because dicts can
+        # do __setitem__ for elements not in the dict and we do not want that.
+        # Note for defaultdict and Counter objects, just calling batch[key] for
+        # with a new keyword will create a new key, value pair in the object.
+        batch[key]
+        batch[key] = value
+
+    # The only acceptable TypeErrors are for an object that doesn't have a __setitem__ or a __getitem__,
+    # which is TypeError("... object does not support item assignment") and TypeError('.. object is not subscriptable')
+    except TypeError as e:
+        if 'object does not support item assignment' in str(e) or 'object is not subscriptable' in str(e):
+            exceptions.append(e)
+            pass
+        else:  # Other type errors should be raised.
+            raise e
+    else:
+        return batch
+
+    try:
+        # Make sure batch has key before setting it.
+        getattr(batch, key)
+        setattr(batch, key, value)
+
+    # If both (setattr or getattr) and __setitem__ raise exceptions then raise both of them.
+    except (AttributeError, TypeError) as e:
+        exceptions.append(e)
+        raise Exception(exceptions)
+    else:
+        return batch
+
+
+def _batch_set_multiple(batch: Any, key: Any, value: Any) -> Any:
+    """Sets multiple key value pairs in a non-tuple batch."""
+    # Numpy arrays and Torch tensors can take tuples and lists as keys, so try to do a normal
+    # __getitem__ call before resulting to list comprehension.
+    try:
+        # Check if one can do a __getitem__ before doing a __setitem__ because dicts can
+        # do __setitem__ for elements not in the dict and we do not want that.
+        batch[key]
+        batch[key] = value
+        return batch
+    # Indexing a list with a sequence results in TypeError
+    # Indexing an array/tensor with a sequence that is longer than the rank of the array
+    # results in an IndexError.
+    except (IndexError, TypeError, KeyError):
+        pass
+    if not hasattr(value, '__len__') or isinstance(value, str):
+        raise ValueError(f'value must be a sequence or array or tensor! and not {type(value)}')
+    if len(key) != len(value):
+        raise ValueError(f'value must be the same length as key ({len(key)}), but it is {len(value)} instead')
+    for single_key, single_value in zip(key, value):
+        batch = _batch_set(batch, single_key, single_value)
```

Comment:
(Same question regarding recursion as the `batch_get`)

```suggestion
        batch = batch_set(batch, single_key, single_value)
```

Reply:
hmmm yeah that could be interesting.

Reply:
but only works for recursively unpacking a key not a batch, so probably too niche of a use case to support.